File: orchestrator/src/storage/redis/scripts/cleanup.lua
-- SPDX-FileCopyrightText: OpenTalk GmbH <mail@opentalk.eu>
--
-- SPDX-License-Identifier: EUPL-1.2

local orchestrator_alive_key = KEYS[1]
local orchestrator_services_key = KEYS[2]
local orchestrators_set_key = KEYS[3]

local orchestrator_id = ARGV[1]

-- Skip if this orchestrator is still alive
if redis.call("EXISTS", orchestrator_alive_key) == 1 then
    return 1 -- still alive
end

-- Check that the given orchestrator ID exists in the global orchestrators set
if redis.call("SISMEMBER", orchestrators_set_key, orchestrator_id) == 0 then
    return 2 -- orchestrator not found
end

-- Clean up every service owned by this orchestrator
local service_ids = redis.call("SMEMBERS", orchestrator_services_key)

for _, service_id in ipairs(service_ids) do
    local instance_key = "ot-orchestrator:service:" .. service_id

    -- Remove resource ownership lookup keys before deleting the resource set
    local resources = redis.call("SMEMBERS", instance_key .. ":resources")
    for _, resource in ipairs(resources) do
        redis.call("DEL", "ot-orchestrator:resource:" .. resource .. ":owner")
    end

    -- Remove the service from its kind set
    local kind = redis.call("HGET", instance_key, "kind")
    if kind then
        redis.call("SREM", "ot-orchestrator:services:" .. kind, service_id)
    end

    -- Remove service instance, metrics, and resource keys
    redis.call("DEL", instance_key)
    redis.call("DEL", instance_key .. ":metrics")
    redis.call("DEL", instance_key .. ":resources")
end

-- Remove orchestrator tracking keys
redis.call("DEL", orchestrator_services_key)
redis.call("SREM", orchestrators_set_key, orchestrator_id)

return 0

File: orchestrator/src/storage/redis/scripts/cleanup.rs
// SPDX-FileCopyrightText: OpenTalk GmbH <mail@opentalk.eu>
//
// SPDX-License-Identifier: EUPL-1.2

use anyhow::Context;
use redis::FromRedisValue;
use uuid::Uuid;

use crate::storage::redis::keys::{
    ORCHESTRATORS_SET_KEY, OrchestratorAliveKey, OrchestratorServicesKey,
};

#[derive(Debug, PartialEq)]
enum ScriptResult {
    /// The orchestrator was dead and its state was removed.
    Ok = 0,
    /// The orchestrator's alive key was still present — no cleanup performed.
    StillAlive = 1,
    /// The orchestrator ID was not found in the global orchestrators set.
    NotFound = 2,
}

impl FromRedisValue for ScriptResult {
    fn from_redis_value(v: redis::Value) -> Result<Self, redis::ParsingError> {
        match i32::from_redis_value(v)? {
            0 => Ok(ScriptResult::Ok),
            1 => Ok(ScriptResult::StillAlive),
            2 => Ok(ScriptResult::NotFound),
            other => Err(format!("unexpected cleanup script return value: {other}").into()),
        }
    }
}

/// Atomically cleans up a single dead orchestrator.
const SCRIPT: &str = include_str!("cleanup.lua");

pub(crate) async fn cleanup(
    client: &redis::Client,
    orchestrator_ids: &[Uuid],
) -> anyhow::Result<Vec<Uuid>> {
    if orchestrator_ids.is_empty() {
        return Ok(Vec::new());
    }

    let script = redis::Script::new(SCRIPT);
    let mut con = client
        .get_multiplexed_async_connection()
        .await
        .context("Failed to get Redis connection for orchestrator cleanup")?;

    let mut cleaned_up = Vec::new();

    for id in orchestrator_ids {
        let id_str = id.clone().to_string();

        let result = script
            .key(OrchestratorAliveKey { id: id_str.clone() })
            .key(OrchestratorServicesKey { id: id_str.clone() })
            .key(ORCHESTRATORS_SET_KEY)
            .arg(id_str.clone())
            .invoke_async::<ScriptResult>(&mut con)
            .await
            .with_context(|| format!("Failed to run cleanup script for orchestrator '{id_str}'"))?;

        match result {
            ScriptResult::Ok => {
                tracing::info!("Cleaned up storage for dead orchestrator '{id_str}'");
                cleaned_up.push(*id);
            }
            ScriptResult::StillAlive => {
                tracing::debug!(
                    "Tried to clean up orchestrator '{id_str}' that is still alive, skipping cleanup"
                );
            }
            ScriptResult::NotFound => {
                // This can happen if the orchestrator was already cleaned up by another instance.
                tracing::debug!(
                    "Tried to clean up orchestrator '{id_str}' that was not found in the global set, skipping cleanup"
                );
            }
        }
    }

    Ok(cleaned_up)
}
